superset/models/core.py

```diff
@@ -314,6 +314,12 @@ def get_sqla_engine(
         params.update(self.get_encrypted_extra())
 
         if DB_CONNECTION_MUTATOR:
+            if not source and request and request.referrer:
```

Comment:
This is similar logic to what was defined in `get_df` however it's been pushed further up the stack. Note we only try to mutate the source if it isn't explicitly defined.